File: GCE/todo-backend/app/routers/tags.py
from fastapi import APIRouter, HTTPException
from datetime import datetime

from app.models import Tag, TagCreate, TagUpdate, TagResponse, TagListResponse
from app.storage import storage

router = APIRouter(prefix="/tags", tags=["Tags"])


@router.get("", response_model=TagListResponse)
def get_all_tags():
    """Get all tags"""
    tags = storage.get_all_tags()
    return TagListResponse(data=tags)


@router.get("/{tag_id}", response_model=TagResponse)
def get_tag(tag_id: str):
    """Get a single tag by ID"""
    tag = storage.get_tag(tag_id)
    if not tag:
        raise HTTPException(
            status_code=404,
            detail={"code": "NOT_FOUND", "message": f"Tag with id '{tag_id}' not found"},
        )
    return TagResponse(data=tag)


@router.post("", response_model=TagResponse, status_code=201)
def create_tag(tag_data: TagCreate):
    """Create a new tag"""
    # Check for duplicate name
    existing_tags = storage.get_all_tags()
    for existing in existing_tags:
        if existing.name.lower() == tag_data.name.lower():
            raise HTTPException(
                status_code=409,
                detail={
                    "code": "CONFLICT",
                    "message": f"Tag with name '{tag_data.name}' already exists",
                },
            )

    tag = Tag(name=tag_data.name, color=tag_data.color)
    created_tag = storage.create_tag(tag)
    return TagResponse(data=created_tag)


@router.patch("/{tag_id}", response_model=TagResponse)
def update_tag(tag_id: str, tag_data: TagUpdate):
    """Update an existing tag"""
    tag = storage.get_tag(tag_id)
    if not tag:
        raise HTTPException(
            status_code=404,
            detail={"code": "NOT_FOUND", "message": f"Tag with id '{tag_id}' not found"},
        )

    # Check for duplicate name if name is being updated
    if tag_data.name is not None:
        existing_tags = storage.get_all_tags()
        for existing in existing_tags:
            if existing.id != tag_id and existing.name.lower() == tag_data.name.lower():
                raise HTTPException(
                    status_code=409,
                    detail={
                        "code": "CONFLICT",
                        "message": f"Tag with name '{tag_data.name}' already exists",
                    },
                )

    tag_dict = tag.model_dump()

    if tag_data.name is not None:
        tag_dict["name"] = tag_data.name
    if tag_data.color is not None:
        tag_dict["color"] = tag_data.color

    tag_dict["updatedAt"] = datetime.utcnow()

    updated_tag = Tag(**tag_dict)
    storage.update_tag(tag_id, updated_tag)
    return TagResponse(data=updated_tag)


@router.delete("/{tag_id}", status_code=204)
def delete_tag(tag_id: str):
    """Delete a tag"""
    if not storage.delete_tag(tag_id):
        raise HTTPException(
            status_code=404,
            detail={"code": "NOT_FOUND", "message": f"Tag with id '{tag_id}' not found"},
        )
    return None

File: GCE/todo-backend/src/routes/tags.js
import { Router } from "express";
import { v4 as uuidv4 } from "uuid";
import * as storage from "../storage.js";

export const tagsRouter = Router();

// GET /api/v1/tags
tagsRouter.get("/", (req, res) => {
  res.json({ data: storage.getAllTags() });
});

// GET /api/v1/tags/:tagId
tagsRouter.get("/:tagId", (req, res) => {
  const tag = storage.getTag(req.params.tagId);
  if (!tag) {
    return res.status(404).json({
      error: {
        code: "NOT_FOUND",
        message: `Tag with id '${req.params.tagId}' not found`,
      },
    });
  }
  res.json({ data: tag });
});

// POST /api/v1/tags
tagsRouter.post("/", (req, res) => {
  const { name, color } = req.body;

  // Validate color format
  if (!/^#[0-9A-Fa-f]{6}$/.test(color)) {
    return res.status(400).json({
      error: {
        code: "VALIDATION_ERROR",
        message: "Color must be in hex format (#RRGGBB)",
      },
    });
  }

  // Check for duplicate name
  const existingTags = storage.getAllTags();
  if (existingTags.some((t) => t.name.toLowerCase() === name.toLowerCase())) {
    return res.status(409).json({
      error: {
        code: "CONFLICT",
        message: `Tag with name '${name}' already exists`,
      },
    });
  }

  const now = new Date().toISOString();
  const tag = { id: uuidv4(), name, color, createdAt: now, updatedAt: now };

  storage.createTag(tag);
  res.status(201).json({ data: tag });
});

// PATCH /api/v1/tags/:tagId
tagsRouter.patch("/:tagId", (req, res) => {
  const tag = storage.getTag(req.params.tagId);
  if (!tag) {
    return res.status(404).json({
      error: {
        code: "NOT_FOUND",
        message: `Tag with id '${req.params.tagId}' not found`,
      },
    });
  }

  const { name, color } = req.body;

  // Validate color format if provided
  if (color && !/^#[0-9A-Fa-f]{6}$/.test(color)) {
    return res.status(400).json({
      error: {
        code: "VALIDATION_ERROR",
        message: "Color must be in hex format (#RRGGBB)",
      },
    });
  }

  // Check for duplicate name if name is being updated
  if (name) {
    const existingTags = storage.getAllTags();
    if (
      existingTags.some(
        (t) =>
          t.id !== req.params.tagId &&
          t.name.toLowerCase() === name.toLowerCase(),
      )
    ) {
      return res.status(409).json({
        error: {
          code: "CONFLICT",
          message: `Tag with name '${name}' already exists`,
        },
      });
    }
  }

  const updatedTag = { ...tag, updatedAt: new Date().toISOString() };
  if (name !== undefined) updatedTag.name = name;
  if (color !== undefined) updatedTag.color = color;

  storage.updateTag(req.params.tagId, updatedTag);
  res.json({ data: updatedTag });
});

// DELETE /api/v1/tags/:tagId
tagsRouter.delete("/:tagId", (req, res) => {
  if (!storage.deleteTag(req.params.tagId)) {
    return res.status(404).json({
      error: {
        code: "NOT_FOUND",
        message: `Tag with id '${req.params.tagId}' not found`,
      },
    });
  }
  res.status(204).send();
});
